json-schema/array.json

```diff
@@ -0,0 +1,639 @@
+                            "type": "string",
+                            "enum": [
+                                "end",
+                                "start"
+                            ],
+                            "description": "Location of the index chunks in the shard."
+                        }
+                    }
+                }
+            }
+        },
+        "zstd_codec": {
+            "type": "object",
+            "required": [
+                "name"
+            ],
+            "properties": {
+                "name": {
+                    "const": "zstd"
+                },
+                "configuration": {
+                    "type": "object",
+                    "properties": {
+                        "level": {
+                            "type": "integer",
+                            "description": "Compression level"
+                        },
+                        "checksum": {
+                            "type": "boolean",
+                            "description": "Whether to include checksum"
+                        }
+                    },
+                    "additionalProperties": false
+                }
+            },
+            "additionalProperties": false
+        },
+        "crc32c_codec": {
+            "type": "object",
+            "required": [
+                "name"
+            ],
+            "properties": {
+                "name": {
+                    "const": "crc32c"
+                }
+            },
+            "additionalProperties": false
+        },
+        "storage_transformer": {
+            "$ref": "#/$defs/extension_object",
+            "description": "Storage transformer configuration"
+        },
+        "extension_object": {
+            "type": "object",
+            "required": [
+                "name"
+            ],
+            "properties": {
+                "name": {
+                    "type": "string",
+                    "pattern": "^[a-z][a-z0-9-_.]+$|^[a-zA-Z][a-zA-Z0-9+.-]*:",
+                    "description": "Extension name (registered name or URI)"
+                },
+                "configuration": {
+                    "type": "object",
+                    "description": "Optional configuration object for the extension"
+                },
+                "must_understand": {
+                    "type": "boolean",
+                    "default": true,
+                    "description": "Whether implementations must understand this extension"
+                }
+            },
+            "additionalProperties": false
+        },
+        "extension_data_type": {
+            "allOf": [
+                {
+                    "$ref": "#/$defs/extension_object"
+                },
+                {
+                    "properties": {
+                        "name": {
+                            "type": "string",
+                            "not": {
+                                "anyOf": [
```

Comment:
This is a common pattern for all the extension objects (`data_type`, `codec`, etc.). We're using the [`oneOf`](https://json-schema.org/understanding-json-schema/reference/combining#oneOf) keyword to ensure that the data type, say, matches *exactly* one data type definition.

To ensure that a data type like `"bool"` doesn't match against both the core bool data type and an extension data type, we need to prohibit extension types from shadowing a core data type.